modules/launcher/src/main/java/org/mule/runtime/module/launcher/application/ApplicationExtensionsManagerConfigurationBuilder.java

```diff
@@ -6,88 +6,93 @@
-        initialiseIfNeeded(extensionManager, muleContext);
+        final ExtensionManagerAdapter extensionManager = createExtensionManager(muleContext);
 
         for (ApplicationPlugin applicationPlugin : applicationPlugins)
         {
-            final ServiceLoader<Describer> describers = ServiceLoader.load(Describer.class, new ClassLoader(null)
+            URL manifestUrl = applicationPlugin.getArtifactClassLoader().findResource("META-INF/" + EXTENSION_MANIFEST_FILE_NAME);
+            if (manifestUrl == null)
             {
-                @Override
-                public Class<?> loadClass(String name) throws ClassNotFoundException
-                {
-                    return applicationPlugin.getArtifactClassLoader().getClassLoader().loadClass(name);
-                }
+                continue;
+            }
 
-                @Override
-                protected URL findResource(String name)
-                {
-                    return applicationPlugin.getArtifactClassLoader().findResource(name);
-                }
+            ExtensionManifest extensionManifest = extensionManager.parseExtensionManifestXml(manifestUrl);
+            extensionManager.registerExtension(extensionManifest, applicationPlugin.getArtifactClassLoader().getClassLoader());
 
-                @Override
-                protected Enumeration<URL> findResources(String name) throws IOException
-                {
-                    return applicationPlugin.getArtifactClassLoader().findResources(name);
-                }
-            });
+            if (LOGGER.isDebugEnabled())
+            {
+                LOGGER.debug("Discovered extension  " + extensionManifest.getName());
+            }
+        }
+    }
 
-            for (Describer describer : describers)
+    private ExtensionManagerAdapter createExtensionManager(MuleContext muleContext) throws InitialisationException
+    {
+        try
+        {
+            ExtensionManager extensionManager = extensionManagerFactory.createExtensionManager(muleContext);
+            if (extensionManager == null)
             {
-                if (logger.isDebugEnabled())
-                {
-                    logger.debug("Discovered extension: " + describer.getClass().getName());
-                }
+                throw new MuleRuntimeException(MessageFactory.createStaticMessage("extensionManagerFactory returned a null instance"));
+            }
 
-                ExtensionDeclarer declarer = describer.describe(new DefaultDescribingContext());
-                final DefaultExtensionFactory extensionFactory = new DefaultExtensionFactory(new SpiServiceRegistry(), applicationPlugin.getArtifactClassLoader().getClassLoader());
-                final RuntimeExtensionModel extensionModel = extensionFactory.createFrom(declarer);
-                extensionManager.registerExtension(extensionModel);
+            if (!(extensionManager instanceof ExtensionManagerAdapter))
```

Comment:
Why must be an instance of ExtensionManagerAdapter? Why the code is depending on something that returns ExtensionManager but only accepts particular implementations?